src/Dependencies/Collections/Internal/SegmentedHashSetEqualityComparer`1.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// NOTE: This code is derived from an implementation originally in dotnet/runtime:
+// https://github.com/dotnet/runtime/blob/v5.0.7/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSetEqualityComparer.cs
+//
+// See the commentary in https://github.com/dotnet/roslyn/pull/50156 for notes on incorporating changes made to the
+// reference implementation.
+
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Collections.Internal
+{
+    /// <summary>Equality comparer for hashsets of hashsets</summary>
+    internal sealed class SegmentedHashSetEqualityComparer<T> : IEqualityComparer<SegmentedHashSet<T>?>
+    {
+        public bool Equals(SegmentedHashSet<T>? x, SegmentedHashSet<T>? y)
+        {
+            // If they're the exact same instance, they're equal.
+            if (ReferenceEquals(x, y))
+            {
+                return true;
+            }
+
+            // They're not both null, so if either is null, they're not equal.
+            if (x == null || y == null)
+            {
+                return false;
+            }
+
+            var defaultComparer = EqualityComparer<T>.Default;
+
+            // If both sets use the same comparer, they're equal if they're the same
+            // size and one is a "subset" of the other.
+            if (SegmentedHashSet<T>.EqualityComparersAreEqual(x, y))
+            {
+                return x.Count == y.Count && y.IsSubsetOfHashSetWithSameComparer(x);
+            }
+
+            // Otherwise, do an O(N^2) match.
+            foreach (var yi in y)
+            {
+                var found = false;
+                foreach (var xi in x)
+                {
+                    if (defaultComparer.Equals(yi, xi))
+                    {
+                        found = true;
+                        break;
+                    }
+                }
+
+                if (!found)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public int GetHashCode(SegmentedHashSet<T>? obj)
+        {
+            var hashCode = 0; // default to 0 for null/empty set
+
+            if (obj != null)
+            {
+                foreach (var t in obj)
+                {
+                    if (t != null)
+                    {
+                        hashCode ^= t.GetHashCode(); // same hashcode as as default comparer
+                    }
+                }
+            }
+
+            return hashCode;
+        }
+
+        // Equals method for the comparer itself.
+        public override bool Equals(object? obj) => obj is SegmentedHashSetEqualityComparer<T>;
```

Comment:
```suggestion
        public override bool Equals([NotNullWhen(true)] object? obj) => obj is SegmentedHashSetEqualityComparer<T>;
```

This was added in dotnet/runtime:

https://github.com/dotnet/runtime/blob/64245ab3966abde394767fbd6531e7edbea2802e/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSetEqualityComparer.cs#L75

Note: requires `using System.Diagnostics.CodeAnalysis;`

Reply:
➡️ This implementation is taken from the link at the top of this file. The 5.0.7 tag does not contain this attribute; it will be added during the update step if/when a newer version is adopted.